diesel/src/pg/query_builder/tablesample.rs

```diff
@@ -0,0 +1,170 @@
+use crate::{
+    expression::{Expression, ValidGrouping},
+    pg::Pg,
+    query_builder::{AsQuery, AstPass, FromClause, QueryFragment, QueryId, SelectStatement},
+    query_source::QuerySource,
+    result::QueryResult,
+    sql_types::{Double, SmallInt},
+    JoinTo, SelectableExpression, Table,
+};
+use std::marker::PhantomData;
+
+#[doc(hidden)]
+pub trait TablesampleMethod: Clone {
+    fn method_name_sql() -> &'static str;
+}
+
+/// Represents a query with a `TABLESAMPLE` clause.
+#[doc(hidden)]
+#[derive(Debug, Clone, Copy)]
+pub struct Tablesample<S, TSM> {
+    pub source: S,
+    pub method: PhantomData<TSM>,
+    pub portion: i16,
+    pub seed: Option<f64>,
+}
+
+impl<S, TSM> QueryId for Tablesample<S, TSM>
+where
+    S: QueryId,
+    TSM: TablesampleMethod,
+{
+    type QueryId = ();
+    const HAS_STATIC_QUERY_ID: bool = false;
+}
+
+impl<S, TSM> QuerySource for Tablesample<S, TSM>
+where
+    S: Table + Clone,
+    TSM: TablesampleMethod,
+    <S as QuerySource>::DefaultSelection:
+        ValidGrouping<()> + SelectableExpression<Tablesample<S, TSM>>,
+{
+    type FromClause = Self;
+    type DefaultSelection = <S as QuerySource>::DefaultSelection;
+
+    fn from_clause(&self) -> Self::FromClause {
+        self.clone()
+    }
+
+    fn default_selection(&self) -> Self::DefaultSelection {
+        self.source.default_selection()
+    }
+}
+
+impl<S, TSM> QueryFragment<Pg> for Tablesample<S, TSM>
+where
+    S: QueryFragment<Pg>,
+    TSM: TablesampleMethod,
+{
+    fn walk_ast<'b>(&'b self, mut out: AstPass<'_, 'b, Pg>) -> QueryResult<()> {
+        self.source.walk_ast(out.reborrow())?;
+        out.push_sql(" TABLESAMPLE ");
+        out.push_sql(TSM::method_name_sql());
+        out.push_sql("(");
+        out.push_bind_param::<SmallInt, _>(&self.portion)?;
+        out.push_sql(")");
+        if let Some(f) = &self.seed {
+            out.push_sql(" REPEATABLE(");
+            out.push_bind_param::<Double, _>(f)?;
+            out.push_sql(")");
+        }
+        Ok(())
+    }
+}
+
+impl<S, TSM> AsQuery for Tablesample<S, TSM>
+where
+    S: Table + Clone,
+    TSM: TablesampleMethod,
+    <S as QuerySource>::DefaultSelection:
+        ValidGrouping<()> + SelectableExpression<Tablesample<S, TSM>>,
+{
+    type SqlType = <<Self as QuerySource>::DefaultSelection as Expression>::SqlType;
+    type Query = SelectStatement<FromClause<Self>>;
+    fn as_query(self) -> Self::Query {
+        SelectStatement::simple(self)
+    }
+}
+
+impl<S, T, TSM> JoinTo<T> for Tablesample<S, TSM>
+where
+    S: JoinTo<T>,
+    T: Table,
+    S: Table,
+    TSM: TablesampleMethod,
+{
+    type FromClause = <S as JoinTo<T>>::FromClause;
+    type OnClause = <S as JoinTo<T>>::OnClause;
+
+    fn join_target(rhs: T) -> (Self::FromClause, Self::OnClause) {
+        <S as JoinTo<T>>::join_target(rhs)
+    }
+}
+
+impl<S, TSM> Table for Tablesample<S, TSM>
+where
+    S: Table + Clone + AsQuery,
```

Comment:
I think it should be fine to just skip these `Clone` bounds, as they are not needed, right?

Reply:
I believe these constraints are needed. Without them we run into issues with the `from_clause` method where the source is cloned. Without them we run into this compile error:

```
error[E0308]: mismatched types
  --> diesel/src/pg/query_builder/tablesample.rs:91:9
   |
90 |     fn from_clause(&self) -> Self::FromClause {
   |                              ---------------- expected `Tablesample<S, TSM>` because of return type
91 |         self.clone()
   |         ^^^^^^^^^^^^ expected `Tablesample<S, TSM>`, found `&Tablesample<S, TSM>`
   |
   = note: expected struct `Tablesample<_, _>`
           found reference `&Tablesample<_, _>`
note: `Tablesample<S, TSM>` does not implement `Clone`, so `&Tablesample<S, TSM>` was cloned instead
  --> diesel/src/pg/query_builder/tablesample.rs:91:9
   |
91 |         self.clone()
   |         ^^^^
   = help: `Clone` is not implemented because the trait bound `S: Clone` is not satisfied
help: consider annotating `Tablesample<S, TSM>` with `#[derive(Clone)]`
   |
38 + #[derive(Clone)]
39 | pub struct Tablesample<S, TSM>
   |

For more information about this error, try `rustc --explain E0308`.
error: could not compile `diesel` (lib) due to 1 previous error
```

Reply:
:+1: